### DISCUSSION TITLE
Dynamic OData with entity Framework

### DISCUSSION THREAD

**Discussion by lorenzocas** [OData/AspNetCoreOData#1317] [Q&A]
Hello,
i'm trying to adapt https://github.com/OData/AspNetCoreOData/tree/main/sample/ODataDynamicModel to use entity framework, but i'm not sure how to implement the ef querying part in the sample. I think i should change the Get() Methods in MyDataSource class, but i'm not sure how. Could you please give me some hints how to do this ?

Thanks
Best regards

**Comment by lorenzocas** [comment]
I'm adding some additiona info to clarify the problem.
I created a project (.Net 8) where I generate an entity framework context dinamically using reflection, based on some metadata (table and columns types definition). 
Then inside OnModelCreating i do something like this:

`modelBuilder.ApplyConfigurationsFromAssembly(Assembly.GetExecutingAssembly());`

Now I need to query this context using odata and I'm trying to modify this sample project: 

https://github.com/OData/AspNetCoreOData/tree/main/sample/ODataDynamicModel

The sample just generate data programmatically like this:

` EdmEntityObject entity = new EdmEntityObject(entityType);
 entity.TrySetPropertyValue("Name", "abc");
 entity.TrySetPropertyValue("ID", 1);
 entity.TrySetPropertyValue("DetailInfo", CreateDetailInfo(88, "abc_detailinfo", entity.ActualEdmType));
 entity.TrySetPropertyValue("ContainedDetailInfo", CreateContainedDetailInfo(88, "abc_containeddetailinfo", entity.ActualEdmType));

 collection.Add(entity);`

In MyDataSource class. I need to modify it to get data directly from my datacontext passing all the odata filters.

Is this possible ?

Thanks

**Reply by Remo** [reply]
I'm a bit confused, why do you even want to use this dynamic approach if you have a data context? If you have proper entities, just register them and you get everything out of the box.

If you already have entities and just want to register them dynamically, implement the magic in your context.
Something like this:

```csharp
protected override void OnModelCreating(ModelBuilder builder)
{
    foreach (var entity in YourEntities)
    {

        builder.Entity(....entity.GetType()...);
    }

    base.OnModelCreating(builder);
}
```

**Comment by lorenzocas** [comment]
How Should I implement the ODataController ?
It should be a dynamic controller where I don't need to specify a method for each model.
If I add a table/columns in the metadata at runtime my Entityframework context is updated and the odata controller should be able to pickup the newly added entity and respond to my rest queries.

**Reply by Remo** [reply]
I haven't tested it, but I guess you would have to use reflection in your controller method. Something like this:

```
[HttpGet]
[EnableQuery]
[ODataRoute("{entityset}")]
public IActionResult Get(string entityset)
{
    var method = Context
        .GetType()
        .GetMethods()
        .First(x => x.IsGenericMethod && x.Name == "Set");

    MethodInfo generic = method.MakeGenericMethod(...GetType..Of.Entity...);
    var set = generic.Invoke(Context, null);

    return Ok(set);
}
```

The context has a Set method with which you can get a dynamically registered DbSet. You'll have to implement a few things from scratch though. A lot like shown in this example https://github.com/OData/AspNetCoreOData/blob/main/sample/ODataDynamicModel/Controllers/HandleAllController.cs

**Comment by lorenzocas** [comment]
It works like a charm (filters included). 
Instead of [ODataRoute("{entityset}")] (not supported in .Net 8) I had to use [HttpGet("odata/{entityset}")].

In the controller:

`[HttpGet("odata/{entityset}")]
[EnableQuery]
//[ODataRoute("{entityset}")]
public IActionResult Get(string entityset)
{
    var method = _context
        .GetType()
        .GetMethods()
        .First(x => x.IsGenericMethod && x.Name == "Set");

    Type t = Type.GetType($"KonvergenceCore.Server.Database.Models.{entityset}");

    MethodInfo generic = method.MakeGenericMethod(t);
    var set = generic.Invoke(_context, null);

    return Ok(set);
}`

In ApplicationEdmModel:

`public static IEdmModel GetEdmModel()
        {
            var modelBuilder = new ODataConventionModelBuilder();

            modelBuilder.Namespace = "WideWorldImportersService";

            var q = from t in Assembly.GetExecutingAssembly().GetTypes()
                    where t.IsClass && t.Namespace == "KonvergenceCore.Server.Database.Models" && Attribute.GetCustomAttribute(t, typeof(CompilerGeneratedAttribute)) == null
                    select t;
            q.ToList().ForEach((t) =>
            {
                var method = modelBuilder.GetType().GetMethod("EntitySet", new Type[] { typeof(string) });

                var genMethos = method.MakeGenericMethod(t);

                object[] args = {
                   t.Name
                };

                genMethos.Invoke(modelBuilder, args);
            });

            //modelBuilder.EntitySet<Person>("Persons");

            // Configure EntityTypes, that could not be mapped using Conventions. We
            // could also add Attributes to the Model, but I want to avoid mixing the
            // EF Core Fluent API and Attributes.
            //modelBuilder.EntityType<Person>().HasKey(s => new { s.PersonId });

            // Build the Spatial Types:
            BuildGeometryTypes(modelBuilder);

            // Send as Lower Camel Case Properties, so the JSON looks better:
            modelBuilder.EnableLowerCamelCase();

            return modelBuilder.GetEdmModel();
        }`

The only problem left is related to this code in the startup.cs ConfigureService:

` .AddOData((opt) =>
 {
     opt.AddRouteComponents(
         routePrefix: "odata",
         model: ApplicationEdmModel.GetEdmModel(), configureServices: svcs =>
         {
             svcs.AddSingleton<ODataBatchHandler>(new DefaultODataBatchHandler());
             svcs.AddSingleton<IFilterBinder, GeoDistanceFilterBinder>();
         })
     .EnableQueryFeatures().Select().Expand().OrderBy().Filter().Count();
 });`

This is executed only once at application startup. So the model is not updated after I add another class/entity to the context at runtime.
Any Ideas ?

Thanks a lot man !!!

**Reply by Remo** [reply]
I think the `EntitySet` method allows you to specify all these options. Something like `modelBuilder.EntityType<Person>().HasKey(s => new { s.PersonId }).Select().Filter();`
You would have to do that on your reflection code, can't do that out of my head, but maybe you can give it a try.
